### PR TITLE
optimize jpg, jpeg, and png images to webp

### PR DESCRIPTION
**Pull request by tekstrand**
## What
Convert all jpg, jpeg, and png images to webp
Screenshot 2024-02-02 at 9 14 15 PM

## Why
No one likes unnecessarily large files

### CODE REVIEW

**Comment by pdxlocations** [comment]
Is webp a good format to use? It's not a standard I'm familiar with.

**Comment by tekstrand** [comment]
Absolutely! https://caniuse.com/webp

It is supported by all major browsers

**Comment by pdxlocations** [comment]
> Absolutely! https://caniuse.com/webp
> 
> It is supported by all major browsers

Is it easy for the community to use for contributing? Not that they have to I suppose. I've just literally never heard of it. Web development is obviously not my profession.

**Comment by pdxlocations** [comment]
Is there a case where this format wouldn't be supported in a future use? Export to PDF for one example.

**Comment by tekstrand** [comment]
No problem. It was new to me a few years ago as well. All major CDN's now use this, often converting on the fly if other formats are uploaded.

As for working with the format - yes, most all image programs support opening, editing, and saving in the webp format. Alternatively, contributors may continue to add other file formats moving forward(jpg, png, etc) they just won't be optimized. 

Eventually we should consider implementing an auto optimization process either via a [git action](Might be nice to automate it though for future images https://css-tricks.com/optimize-images-with-a-github-action/) or a plugin like [plugin-ideal-image](https://docusaurus.io/docs/api/plugins/@docusaurus/plugin-ideal-image) for our documentation platform.

RE: Case where format wouldn't be supported: Can you tell me more about what "Export to PDF" would be used for? Are we hoping to be able to save the documentation as a PDF? Something else?

**Comment by pdxlocations** [comment]
> RE: Case where format wouldn't be supported: Can you tell me more about what "Export to PDF" would be used for? Are we hoping to be able to save the documentation as a PDF? Something else?

Yes, this feature has been requested numerous times.  Since the tech is for off-grid use, lots of folks want a hard copy to store in their go bag.  It's one thing we're hoping the "next platform" would automate.

**Comment by pdxlocations** [comment]
Not sure I'm doing this right, or if it matters, but it seems github can't preview these files in-line?

https://github.com/meshtastic/meshtastic/blob/efcb2e97c9b23c2b053776e28abc72de35a35c4a/static/img/configuration/max-hops.webp
vs
https://github.com/meshtastic/meshtastic/blob/5b7d98fb894e0ef3b754a04764c81797c8fb2ba5/static/img/configuration/max-hops.png

**Comment by tekstrand** [comment]
It may not show "inline" but does show in raw https://raw.githubusercontent.com/meshtastic/meshtastic/efcb2e97c9b23c2b053776e28abc72de35a35c4a/static/img/configuration/max-hops.webp

**Comment by pdxlocations** [comment]
> RE: Case where format wouldn't be supported: Can you tell me more about what "Export to PDF" would be used for? Are we hoping to be able to save the documentation as a PDF? Something else?

https://github.com/meshtastic/meshtastic/issues/226

**Comment by tekstrand** [comment]
As for pdf, I'm on mobile now so links are a bit more challenging but I see a number of good options out there right now that will allow us to automatically generate a PDF of the Docusaurus website using a GitHub action. I can take a look into that this weekend. Webp format images will not have any effect on this no matter the direction we go in the future.

**Comment by pdxlocations** [comment]
> As for pdf, I'm on mobile now so links are a bit more challenging but I see a number of good options out there right now that will allow us to automatically generate a PDF of the Docusaurus website using a GitHub action. I can take a look into that this weekend. Webp format images will not have any effect on this no matter the direction we go in the future.



> I see a number of good options out there right now that will allow us to automatically generate a PDF of the Docusaurus website using a GitHub action.

It was our use of tabs that made this more difficult that it should be.

**Comment by tekstrand** [comment]
Oh right...the tabs. Those things really are bad for a lot of reasons.

After thinking more webp would actually be a prerequisite for pdf because the file size without it would be terrible.

I'll still take a look at PDF this weekend and see if I can come up with something.

File: docs/configuration/module/serial.mdx [hunk 1]
```diff
@@ -11,7 +11,7 @@ The serial module config options are: Enabled, Echo, Mode, Receive GPIO, Transmi
 
 This is an interface to talk to and control your Meshtastic device over a serial port. The module can be set to different [modes](#mode), each fulfilling a different use-case.
 
-![image](https://user-images.githubusercontent.com/9000580/205529843-962c3187-8411-452c-b729-42c58b1571f5.png)
+![image](https://user-images.githubusercontent.com/9000580/205529843-962c3187-8411-452c-b729-42c58b1571f5.webp)
```
Comment (pdxlocations) [inline docs/configuration/module/serial.mdx:14]: This links out of Meshtastic.  A build attempt may find them all.

File: docs/getting-started/flashing-firmware/esp32/external-serial-adapter.mdx [hunk 2]
```diff
@@ -33,7 +33,7 @@ The adapter featured in this document can be purchased from:
 https://www.amazon.com/gp/product/B078W5L8W1/
 
 Plug the adapter into your computer without connecting it to any devices yet. Ensure that your computer has drivers installed for the adapter. On Windows, the correct drivers should install automatically.
-![image](https://user-images.githubusercontent.com/9000580/168445869-f73667f5-0169-46fd-ae13-15c3abad31aa.png)
+![image](https://user-images.githubusercontent.com/9000580/168445869-f73667f5-0169-46fd-ae13-15c3abad31aa.webp)
```
Comment (pdxlocations) [inline docs/getting-started/flashing-firmware/esp32/external-serial-adapter.mdx:36]: flag

File: docs/getting-started/flashing-firmware/esp32/external-serial-adapter.mdx [hunk 3]
```diff
@@ -48,7 +48,7 @@ Disconnect your USB to Serial Adapter from the computer before starting this pro
 4. Connect either the 5V pin of the adapter to the 5V pin of the device (illustrated) or the 3.3V pin of the adapter to the 3.3V pin of device.
 5. Bridge GPIO 0 to GND on the device with a jumper. (This places the device into flash mode when the device is powered up)
    Example wiring featuring a T-Beam
-   ![image](https://user-images.githubusercontent.com/9000580/168446438-d14f34a0-b92d-4716-87fd-e0c2927a35fd.png)
+   ![image](https://user-images.githubusercontent.com/9000580/168446438-d14f34a0-b92d-4716-87fd-e0c2927a35fd.webp)
```
Comment (pdxlocations) [inline docs/getting-started/flashing-firmware/esp32/external-serial-adapter.mdx:51]: flag

File: docs/getting-started/flashing-firmware/esp32/external-serial-adapter.mdx [hunk 4]
```diff
@@ -59,7 +59,7 @@ After following the steps above, your device should be in flash mode. You can fl
 
 After flashing the device is complete, reset your device (via the RST button if available).
 If you have the Meshtastic Python CLI installed, you can run `meshtastic --noproto` to connect the device again over the adapter and view the serial output to confirm Meshtastic installed correctly.
-![image](https://user-images.githubusercontent.com/9000580/168447159-71a6546a-f487-4bc2-86c1-4c489b2a8975.png)
+![image](https://user-images.githubusercontent.com/9000580/168447159-71a6546a-f487-4bc2-86c1-4c489b2a8975.webp)
```
Comment (pdxlocations) [inline docs/getting-started/flashing-firmware/esp32/external-serial-adapter.mdx:62]: flag

File: docs/hardware/devices/heltec/index.mdx [hunk 5]
```diff
@@ -83,9 +83,9 @@ This device may have issues charging a connected battery if utilizing a USB-C to
 
 ### Pin Map
 
-![HTIT-WSL_V3_PIN_MAP](</img/hardware/HTIT-WB32LA(F)_V3.png>)
+![HTIT-WSL_V3_PIN_MAP](</img/hardware/HTIT-WB32LA(F)_V3.webp>)
 
-Image Source: [Heltec](<https://resource.heltec.cn/download/WiFi_LoRa32_V3/HTIT-WB32LA(F)_V3.png>)
+Image Source: [Heltec](<https://resource.heltec.cn/download/WiFi_LoRa32_V3/HTIT-WB32LA(F)_V3.webp>)
```
Comment (pdxlocations) [inline docs/hardware/devices/heltec/index.mdx:88]: flag

File: docs/hardware/devices/heltec/index.mdx [hunk 6]
```diff
@@ -135,7 +135,7 @@ This device may have issues charging a connected battery if utilizing a USB-C to
 
 ![HTIT-WSL_V3_PIN_MAP](/img/hardware/heltek-wsl-v3.svg)
 
-Image Source: [Heltec](https://resource.heltec.cn/download/Wireless_Stick_Lite_V3/HTIT-WSL_V3.png)
+Image Source: [Heltec](https://resource.heltec.cn/download/Wireless_Stick_Lite_V3/HTIT-WSL_V3.webp)
```
Comment (pdxlocations) [inline docs/hardware/devices/heltec/index.mdx:138]: flag

**Review by pdxlocations — CHANGES_REQUESTED** [review]

File: docs/configuration/module/serial.mdx (hunk 1 above)
Comment (pdxlocations) [reply docs/configuration/module/serial.mdx:14]: Build may not find missing images but it does find:
  Exhaustive list of all broken links found:
  - Broken link on source page path = /docs/software/integrations/mqtt/nodered:
     -> linking to /documents/mqtt/NodeRedTwo.webp
     -> linking to /documents/mqtt/NodeRedThree.webp
     -> linking to /documents/mqtt/NR_nodes.webp
     -> linking to /documents/mqtt/Broker1.webp
     -> linking to /documents/mqtt/Consume.webp
     -> linking to /documents/mqtt/Inject.webp
     -> linking to /documents/mqtt/Forward.webp
     -> linking to /documents/mqtt/DecodeNewest.webp
     -> linking to /documents/mqtt/Mapping.webp
     -> linking to /documents/mqtt/EncodingPosition.webp
     -> linking to /documents/mqtt/PosJSON.webp

Comment (tekstrand) [reply docs/configuration/module/serial.mdx:14]: 🤦 not sure how I missed this. Didn't see the build errors I guess. Thanks

Comment (tekstrand) [reply docs/configuration/module/serial.mdx:14]: Do you know why these images are under `documents` instead of `static/img`?

**Comment by tekstrand** [comment]
Okay thank you for catching that, my local build wasn't failing with those. This time I used regex to update paths
`(img/[^"'\s]+?)\.(jpg|jpeg|png)` <- this search pattern makes sure that the filename is prefixed with an `img` folder in the name, which worked for us as none of the externally sourced images had `img` in the path.

I'd still like to figure out why those specific images aren't in `static/img` and maybe move them in a future PR.

**Comment by tekstrand** [comment]
Screenshot 2024-02-03 at 7 46 14 AM

File: docs/configuration/module/serial.mdx (hunk 1 above)
Comment (pdxlocations) [reply docs/configuration/module/serial.mdx]: that page is pretty old, i'm sure it'd be safe to move em.

**Review by sachaw — APPROVED** [review]

**Review by pdxlocations — APPROVED** [review]

**Pull request merged**